src/chainstate/stacks/transaction.rs

```diff
@@ -17,137 +17,3215 @@
+                }
+            }
+        }
+
+        let next_sighash = self.tx.sign_next_origin(&self.sighash, privk)?;
+        self.sighash = next_sighash;
+        Ok(())
+    }
+
+    pub fn append_origin(&mut self, pubk: &StacksPublicKey) -> Result<(), net_error> {
+        if self.check_overlap && self.origin_done {
+            // can't append another origin key
+            return Err(net_error::SigningError("Cannot append public key to origin after sponsor key".to_string()));
+        }
+
+        self.tx.append_next_origin(pubk)
+    }
+    
+    pub fn sign_sponsor(&mut self, privk: &StacksPrivateKey) -> Result<(), net_error> {
+        match self.tx.auth {
+            TransactionAuth::Sponsored(_, ref sponsor_condition) => {
+                if self.check_oversign && sponsor_condition.num_signatures() >= sponsor_condition.signatures_required() {
+                    return Err(net_error::SigningError("Sponsor would have too many signatures".to_string()));
+                }
+            },
+            _ => {}
+        }
+
+        let next_sighash = self.tx.sign_next_sponsor(&self.sighash, privk)?;
+        self.sighash = next_sighash;
+        self.origin_done = true;
+        Ok(())
+    }
+
+    pub fn append_sponsor(&mut self, pubk: &StacksPublicKey) -> Result<(), net_error> {
+        self.tx.append_next_sponsor(pubk)
+    }
+
+    pub fn pop_origin_auth_field(&mut self) -> Option<TransactionAuthField> {
+        match self.tx.auth {
+            TransactionAuth::Standard(ref mut origin_condition) => StacksTransaction::pop_auth_field(origin_condition),
+            TransactionAuth::Sponsored(ref mut origin_condition, _) => StacksTransaction::pop_auth_field(origin_condition)
+        }
+    }
+
+    pub fn pop_sponsor_auth_field(&mut self) -> Option<TransactionAuthField> {
+        match self.tx.auth {
+            TransactionAuth::Sponsored(_, ref mut sponsor_condition) => StacksTransaction::pop_auth_field(sponsor_condition),
+            _ => None
+        }
+    }
+
+    pub fn complete(&self) -> bool {
+        match self.tx.auth {
+            TransactionAuth::Standard(ref origin_condition) => {
+                origin_condition.num_signatures() >= origin_condition.signatures_required()
+            },
+            TransactionAuth::Sponsored(ref origin_condition, ref sponsored_condition) => {
+                origin_condition.num_signatures() >= origin_condition.signatures_required() &&
+                sponsored_condition.num_signatures() >= sponsored_condition.signatures_required() &&
+                (self.origin_done || !self.check_overlap)
+            }
+        }
+    }
+
+    pub fn get_tx_incomplete(&self) -> StacksTransaction {
+        self.tx.clone()
+    }
+
+    pub fn get_tx(&self) -> Option<StacksTransaction> {
+        if self.complete() {
+            Some(self.tx.clone())
+        }
+        else {
+            None
+        }
+    }
+
+    pub fn get_incomplete_tx(&self) -> StacksTransaction {
```

Comment:
This is a duplicate of `get_tx_incomplete`

Reply:
Lol how'd that happen?  Removed.